internal/controller/nonadmindownloadrequest_controller.go

```diff
@@ -0,0 +1,299 @@
+		veleroDR.Status.Phase == velerov1.DownloadRequestPhaseProcessed &&
+		!veleroDR.Status.Expiration.Before(&metav1.Time{Time: time.Now()}) {
+		// copy req, prepare to patch
+		prePatch := req.DeepCopy()
+		// copy status to NADR from VDR
+		req.Status.VeleroDownloadRequest.Status = &veleroDR.Status
+		req.Status.Phase = nacv1alpha1.NonAdminPhaseCreated
+		// veleroDR is processed, update NADR status
+		// clear conditions
+		req.Status.Conditions = []metav1.Condition{}
+		if patchErr := r.Status().Patch(ctx, req, client.MergeFrom(prePatch)); patchErr != nil {
+			logger.Error(patchErr, "unable to patch status", req.Kind, req.Name)
+			return patchErr
+		}
+	}
+	return nil
+}
+
+// processDownloadRequest creates download requests after performing NaBSL checks on NAB/NAR from the downloadRequest
+func (r *NonAdminDownloadRequestReconciler) processDownloadRequest(ctx context.Context, veleroDR *velerov1.DownloadRequest, req *nacv1alpha1.NonAdminDownloadRequest) error {
+	logger := log.FromContext(ctx)
+	var nab nacv1alpha1.NonAdminBackup // holds nonadminbackup
+	var nabName string                 // holds nonadminbackup name
+	switch req.Spec.Target.Kind {
+	case velerov1.DownloadTargetKindBackupLog,
+		velerov1.DownloadTargetKindBackupContents,
+		velerov1.DownloadTargetKindBackupVolumeSnapshots,
+		velerov1.DownloadTargetKindBackupItemOperations,
+		velerov1.DownloadTargetKindBackupResourceList,
+		velerov1.DownloadTargetKindBackupResults,
+		velerov1.DownloadTargetKindCSIBackupVolumeSnapshots,
+		velerov1.DownloadTargetKindCSIBackupVolumeSnapshotContents,
+		velerov1.DownloadTargetKindBackupVolumeInfos:
+		nabName = req.Spec.Target.Name
+	case velerov1.DownloadTargetKindRestoreLog,
+		velerov1.DownloadTargetKindRestoreResults,
+		velerov1.DownloadTargetKindRestoreResourceList,
+		velerov1.DownloadTargetKindRestoreItemOperations,
+		velerov1.DownloadTargetKindRestoreVolumeInfo:
+		var nar nacv1alpha1.NonAdminRestore
+		if err := r.Get(ctx, types.NamespacedName{Namespace: req.Namespace, Name: req.Spec.Target.Name}, &nar); err != nil {
+			patchErr := r.patchAddStatusConditionTypeTrueBackoff(ctx, req, nacv1alpha1.ConditionNonAdminRestoreNotAvailable, "cannot get nonadminrestore")
+			logger.Error(patchErr, statusPatchErr, req.Kind, req.Name)
+			logger.Error(err, "unable to get nar", nar.Name, nar.Namespace)
+			return err
+		}
+		// check nar for nabsl
+		nabName = nar.NonAdminBackupName()
+		veleroDR.Spec.Target.Name = nar.VeleroRestoreName()
+	}
+	if err := r.Get(ctx, types.NamespacedName{Namespace: req.Namespace, Name: nabName}, &nab); err != nil {
+		patchErr := r.patchAddStatusConditionTypeTrueBackoff(ctx, req, nacv1alpha1.ConditionNonAdminBackupNotAvailable, "cannot get nonadminbackup")
+		logger.Error(patchErr, statusPatchErr, req.Kind, req.Name)
+		logger.Error(err, "unable to get nab", nab.Name, nab.Namespace)
+		return err
+	}
+	// error if nab does not use nabsl
+	if !nab.UsesNaBSL() {
```

Comment:
Should we do this in a validation function ?

Reply:
If you can run validation function in the middle perhaps.. otherwise validation func will have to perform duplicate get calls. So I think this is fine.